src/conn.c

```diff
@@ -18,34 +19,45 @@ static int init_read(struct conn *c, uv_buf_t *buf, size_t size)
 }
 
 static int read_message(struct conn *c);
-static void conn_write_cb(struct transport *transport, int status)
+static void conn_write_cb(uv_write_t *req, int status)
 {
-	struct conn *c = transport->data;
+	struct transport *t = req->data;
+	assert(t != NULL);
+	struct conn *c = t->data;
+	assert(c != NULL);
 	bool finished;
 	int rv;
 	if (status != 0) {
 		tracef("write cb status %d", status);
 		goto abort;
 	}
+	if (c->closed) {
+		tracef("connection closing");
+		goto abort;
+	}
```

Comment:
what happened before when we didn't check this field?

Reply:
```
2024-02-05T13:22:14.7011433Z LIBDQLITE[031224] 2024-02-05T13:20:24.448881817 031224 DEBUG   clientSendInterrupt  src/client/protocol.c:842 client send interrupt
2024-02-05T13:22:14.7011948Z LIBDQLITE[031224] 2024-02-05T13:20:24.448907059 031225 DEBUG   conn_write_cb        src/conn.c:39  request finished: 0
2024-02-05T13:22:14.7012473Z LIBDQLITE[031224] 2024-02-05T13:20:24.448915044 031225 DEBUG   gateway__handle      src/gateway.c:1389 gateway handle
2024-02-05T13:22:14.7012987Z LIBDQLITE[031224] 2024-02-05T13:20:24.448920774 031225 DEBUG   handle_interrupt     src/gateway.c:953 handle interrupt
2024-02-05T13:22:14.7013454Z LIBDQLITE[031224] 2024-02-05T13:20:24.448934319 031225 DEBUG   conn_write_cb        src/conn.c:39  request finished: 1
2024-02-05T13:22:14.7013922Z LIBDQLITE[031224] 2024-02-05T13:20:24.448948366 031225 DEBUG   conn_write_cb        src/conn.c:39  request finished: 1
2024-02-05T13:22:14.7014652Z LIBDQLITE[031224] 2024-02-05T13:20:24.448959501 031224 DEBUG   clientSendQuery      src/client/protocol.c:663 client send query stmt_id 4
2024-02-05T13:22:14.7015188Z LIBDQLITE[031224] 2024-02-05T13:20:24.448969540 031224 DEBUG   clientRecvRows       src/client/protocol.c:706 client recv rows
2024-02-05T13:22:14.7015659Z LIBDQLITE[031224] 2024-02-05T13:20:24.448983120 031225 DEBUG   gateway__handle      src/gateway.c:1389 gateway handle
2024-02-05T13:22:14.7016149Z LIBDQLITE[031224] 2024-02-05T13:20:24.448992538 031225 DEBUG   handle_query         src/gateway.c:591 handle query schema:1
2024-02-05T13:22:14.7016613Z LIBDQLITE[031224] 2024-02-05T13:20:24.449002506 031225 DEBUG   leader__barrier      src/leader.c:466 leader barrier
2024-02-05T13:22:14.7017053Z LIBDQLITE[031224] 2024-02-05T13:20:24.449006524 031225 DEBUG   leader__barrier      src/leader.c:469 not needed
2024-02-05T13:22:14.7017568Z LIBDQLITE[031224] 2024-02-05T13:20:24.449010161 031225 DEBUG   query_barrier_cb     src/gateway.c:553 query barrier cb status:0
2024-02-05T13:22:14.7018061Z LIBDQLITE[031224] 2024-02-05T13:20:24.449193808 031224 DEBUG   clientClose          src/client/protocol.c:332 client close
2024-02-05T13:22:14.7018528Z LIBDQLITE[031224] 2024-02-05T13:20:24.449205194 031225 DEBUG   conn_write_cb        src/conn.c:39  request finished: 0
2024-02-05T13:22:14.7019020Z LIBDQLITE[031224] 2024-02-05T13:20:24.449213469 031225 DEBUG   read_message_cb      src/conn.c:205 read error -104
2024-02-05T13:22:14.7019453Z LIBDQLITE[031224] 2024-02-05T13:20:24.449219020 031225 DEBUG   conn__stop           src/conn.c:353 conn stop
2024-02-05T13:22:14.7019911Z LIBDQLITE[031224] 2024-02-05T13:20:24.449224450 031225 DEBUG   gateway__close       src/gateway.c:86  gateway close
2024-02-05T13:22:14.7020385Z LIBDQLITE[031224] 2024-02-05T13:20:24.449228893 031224 DEBUG   dqlite_node_stop     src/server.c:871 dqlite node stop
2024-02-05T13:22:14.7020833Z LIBDQLITE[031224] 2024-02-05T13:20:24.449239658 031225 DEBUG   leader__close        src/leader.c:145 leader close
2024-02-05T13:22:14.7020949Z AddressSanitizer:DEADLYSIGNAL
2024-02-05T13:22:14.7021079Z =================================================================
2024-02-05T13:22:14.7021540Z ==31224==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000018 (pc 0x7f95c23daabf bp 0x000000000000 sp 0x7f95beafe060 T1)
2024-02-05T13:22:14.7021705Z ==31224==The signal is caused by a READ memory access.
2024-02-05T13:22:14.7021844Z ==31224==Hint: address points to the zero page.
2024-02-05T13:22:14.7022128Z     #0 0x7f95c23daabe  (/lib/x86_64-linux-gnu/libsqlite3.so.0+0x48abe)
2024-02-05T13:22:14.7022284Z     #1 0x7f95c22d89e4 in query__batch src/query.c:118
2024-02-05T13:22:14.7022431Z     #2 0x7f95c22bac25 in query_batch src/gateway.c:527
2024-02-05T13:22:14.7022605Z     #3 0x7f95c22cc2d6 in gateway__resume src/gateway.c:1448
2024-02-05T13:22:14.7022747Z     #4 0x7f95c229c2df in conn_write_cb src/conn.c:38
2024-02-05T13:22:14.7022997Z     #5 0x7f95c25283af  (/lib/x86_64-linux-gnu/libuv.so.1+0x1b3af)
2024-02-05T13:22:14.7023338Z     #6 0x7f95c2528a30 in uv__stream_destroy (/lib/x86_64-linux-gnu/libuv.so.1+0x1ba30)
2024-02-05T13:22:14.7023616Z     #7 0x7f95c251d8b7 in uv_run (/lib/x86_64-linux-gnu/libuv.so.1+0x108b7)
2024-02-05T13:22:14.7023811Z     #8 0x7f95c22e65fd in taskRun src/server.c:720
2024-02-05T13:22:14.7023997Z     #9 0x7f95c22e6874 in taskStart src/server.c:766
2024-02-05T13:22:14.7024372Z     #10 0x7f95c24c3608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608)
2024-02-05T13:22:14.7024699Z     #11 0x7f95c2144352 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f352)
```

`gateway_resume` tries to continue to step a sqlite stmt while the statement has been freed already. Basically it looks like `g->req` is not properly cleaned up when the gateway is closed, but the close/cleanup code is a mess, is tricky and should be reworked more thoroughly. I fixed this particular issue it in the connection because it did the job and was simpler.

Reply:
Yeah the gateway code needs some love...